### PR TITLE
Geometry 29 2 sven

### PR DESCRIPTION
**Pull request by svenrathgeber**
Hi guys, 
this is my first version for the issue. 
I would be happy, if you could provide some feedback.

Cheers. Sven

### CODE REVIEW

**Comment by svenrathgeber** [comment]
how can I run the test coverage locally ?

**Comment by aherbert** [comment]
JaCoCo is used to collate coverage and is enabled by default. Start by running all the tests or a test of choice (using a regex) then generate a report:

```
mvn test
mvn test -Dtest=MyClassTest
mvn test -Dtest=SomeRegex*Test
mvn test -Dtest=CommonName*
mvn jacoco:report
```

open `[module]/target/site/jacoco/index.html` in a browser.

You can modify code, repeat just your test of choice and rerun the JaCoCo report. It will update using the latest execution data.

**Comment by svenrathgeber** [comment]
@aherbert thanks. I thought, I'd have to use coveralls.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 1]
```diff
@@ -442,14 +593,51 @@ public double getOffset(final Vector3D point) {
...
     @Override
     public boolean sameOrientationAs(final Hyperplane<Vector3D> other) {
         return (((Plane) other).w).dot(w) > 0.0;
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public String toString() {
+        return "Plane [originOffset=" + originOffset + ", projectedOrigin=" + projectedOrigin + ", u=" + u + ", v=" + v + ", w=" + w
+                + ", precision=" + precision + "]";
```
Comment (darkma773r) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: I would trim down the info here and just print out the most useful bits, namely `origin`, `u`, `v`, and `w`. Also, use the name `origin` instead of `projectedOrigin`, since the former is the name of the property as exposed to the outside world (via `getOrigin()`).

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 2]
```diff
@@ -16,174 +16,193 @@
...
-        throws IllegalArgumentException {
-        setNormal(normal);
-        this.precision = precision;
-        this.originOffset = -p.dot(w);
-        setFrame();
+    public static Plane fromPointAndPlaneVectors (Vector3D p, final Vector3D u, final Vector3D v, final DoublePrecisionContext precision)
+    {
+        Vector3D u_norm = u.normalize();
+        Vector3D v_norm = v.normalize();
+        Vector3D w_norm = u_norm.cross(v_norm).normalize();
```
Comment (darkma773r) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: Are we handling the case where the given `u` and `v` vectors are not orthogonal? Take a look at [`QuaternionRotation.createBasisRotation()`](https://github.com/darkma773r/commons-geometry/blob/master/commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/rotation/QuaternionRotation.java#L663) where I do something very similar to this. The general idea is to take the first vector (`u` in this case) as the main anchor vector and normalize it. `v` is then made orthogonal to `u` and normalized at the same time with the `orthogonal` method. The cross product is then used to get the third vector.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 3]
```diff
@@ -16,174 +16,193 @@
...
+    /** Normalized plane normal. */
+    private final Vector3D w;
 
-    /** Third vector of the plane frame (plane normal). */
-    private Vector3D w;
+    /** Offset of the origin with respect to the plane. */
+    private final double originOffset;
 
+    /** orthogonal projection of the 3D-space origin in the plane. */
+    private final Vector3D projectedOrigin;
```
Comment (darkma773r) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: Do we need to store this value? It is easily derived from `w` and `originOffset` and is not used in any internal calculations, so I am leaning toward calculating it directly in the `getOrigin()` method.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 4]
```diff
@@ -16,174 +16,193 @@
...
-    public Plane(final Vector3D normal, final DoublePrecisionContext precision)
-        throws IllegalArgumentException {
-        setNormal(normal);
+    private Plane(Vector3D u, Vector3D v, Vector3D w, Vector3D projectedOrigin, double originOffset,
+            DoublePrecisionContext precision) {
+        this.u = u;
+        this.v = v;
+        this.w = w;
+        
+        if (!areVectorsNormalized(u, v, w, precision))
```
Comment (darkma773r) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: This assertion should not be needed. This constructor is only called internally and the external-facing methods already normalize the inputs, which should throw exceptions on failure. I would add unit tests that pass zero vectors to all public factory methods to verify that the correct exceptions are thrown and then remove this.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 5]
```diff
@@ -16,174 +16,193 @@
...
      */
-    public void reset(final Plane original) {
-        originOffset = original.originOffset;
-        origin       = original.origin;
-        u            = original.u;
-        v            = original.v;
-        w            = original.w;
+    public static Plane fromPoints(final Vector3D p1, final Vector3D p2, final Vector3D p3,
+            final DoublePrecisionContext precision) {
+        return Plane.fromPointAndNormal(p1, p2.subtract(p1).cross(p3.subtract(p1)), precision);
```
Comment (darkma773r) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: Rather than calling `fromPointAndNormal()` to construct the Plane here, why not call `fromPointandPlaneVectors`? That way, the `u` vector will be oriented to point from `p1` to `p2`, giving the caller more control over the orientation of the plane by his/her choice of points.

[the remark this answers is not in the document]
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 6]
```diff
@@ -16,174 +16,193 @@
...
+     * </p>
+     * 
+     * @param plane plane to copy
+     * @return a new plane
      */
-    private void setNormal(final Vector3D normal) {
-        final double norm = Vectors.checkedNorm(normal);
-
-        w = Vector3D.linearCombination(1.0 / norm, normal);
+    public static Plane of(final Plane plane) {
```
[the remark this answers is not in the document]
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 7]
```diff
@@ -16,174 +16,193 @@
...
+     * 
+     * @return a new hyperplane, copy of the instance
      */
-    private void setFrame() {
-        origin = Vector3D.linearCombination(-originOffset, w);
-        u = w.orthogonal();
-        v = w.cross(u);
+    @Deprecated
+    @Override
+    public Plane copySelf() {
```
[the remark this answers is not in the document]
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 8]
```diff
@@ -16,174 +16,193 @@
...
      */
-    public Plane(final Vector3D p, final Vector3D normal, final DoublePrecisionContext precision)
-        throws IllegalArgumentException {
-        setNormal(normal);
-        this.precision = precision;
-        this.originOffset = -p.dot(w);
-        setFrame();
+    public static Plane fromPointAndPlaneVectors (Vector3D p, final Vector3D u, final Vector3D v, final DoublePrecisionContext precision)
+    {
+        Vector3D u_norm = u.normalize();
```
[the remark this answers is not in the document]
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (darkma773r) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: We may have miscommunicated on this. I was picturing that we don't need the `Plane.of(Plane)` method at all. Since `Plane` instances are immutable there's really no point to making a copy of one. I would keep the `final` modifier on parameters throughout, though.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Line.java [hunk 9]
```diff
@@ -236,5 +238,29 @@ public Vector3D intersection(final Line line) {
     public SubLine wholeLine() {
         return new SubLine(this, new IntervalsSet(precision));
     }
+    
+    
 
+    @Override
+    public int hashCode() {
+        throw new IllegalStateException("Must not be used in maps.");
```
Comment (darkma773r) [inline commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Line.java]: We should provide an actual implementation of this method, especially since `equals()` is overridden.

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java [hunk 10]
```diff
@@ -192,50 +211,104 @@ public Vector3D getU() {
...
 
+    /**
+     * 3D line projected onto plane
+     * @param line the line to project
+     * @return the projection of the given line onto the plane.
+     */
+    public Line project(Line line)
+    {
+        Vector3D direction = line.getDirection();
+        Vector3D projection = w.multiply(direction.dot(w)).multiply(1/Math.pow(w.norm(), 2));
```
[the remark this answers is not in the document]
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java (hunk 1 above)
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java (hunk 4 above)
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java (hunk 5 above)
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: I'm struggeling with this one. Some PolyhedronsSetTest fail, when I make the change ... I have to investigate this

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Line.java (hunk 9 above)
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Line.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java (hunk 3 above)
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: I'am undecided. Depends on the way it is used, could be a performance gain or loss. 
Some use cases and profile runs would be great

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java (hunk 2 above)
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java (hunk 5 above)
Comment (svenrathgeber) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: done

File: commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java (hunk 3 above)
Comment (darkma773r) [reply commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/threed/Plane.java]: It is something of a toss-up. The reason I'm leaning toward not storing it is because it is not involved in any of the calculations in the `Hyperplane` methods and therefore won't be used in any of the partitioning logic, which is where this class will get a real workout. That's the approach I took with the 2D `Line` class.
